File: arena-junit/src/main/java/arena/junit/lifecycle/LifecycleLog.java
package arena.junit.lifecycle;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class LifecycleLog {
  public static final String ARENA_ROOT_LOGGER_NAME = "arena";

  private LifecycleLog() {}

  public static String arenaLoggerName(String arenaId) {
    String segment = arenaId == null ? "" : arenaId.trim().replace('.', '_');
    if (segment.isEmpty()) {
      return ARENA_ROOT_LOGGER_NAME;
    }
    return ARENA_ROOT_LOGGER_NAME + "." + segment;
  }

  public static void logTransition(ArenaState state) {
    Logger lg = LoggerFactory.getLogger(arenaLoggerName(state.id));
    int faultCount = state.faults.size();
    String line = faultCount > 0 ? state.state + " | faults=" + faultCount : state.state;
    if (state.isFaulted()) {
      lg.error(line);
    } else {
      lg.info(line);
    }
  }

  public static void logTransitionDocument(String document) {
    ArenaState state;
    try {
      state = ArenaState.parse(document);
    } catch (IllegalArgumentException e) {
      LoggerFactory.getLogger(ARENA_ROOT_LOGGER_NAME)
          .warn("unparseable arena state transition: {}", document);
      return;
    }
    logTransition(state);
  }

  public static void logClosingSummary(ArenaState state) {
    LoggerFactory.getLogger(arenaLoggerName(state.id))
        .info("closing summary | state={}, faults={}", state.state, state.faults.size());
  }

  public static void logClosingSummaryDocument(String document) {
    ArenaState state;
    try {
      state = ArenaState.parse(document);
    } catch (IllegalArgumentException e) {
      LoggerFactory.getLogger(ARENA_ROOT_LOGGER_NAME)
          .warn("unparseable arena closing state: {}", document);
      return;
    }
    logClosingSummary(state);
  }
}

File: arena-xunit/src/Lifecycle/LifecycleLog.cs
using System;
using ArenaDotnet.Xunit.Ffi;
using Microsoft.Extensions.Logging;

namespace ArenaDotnet.Xunit.Lifecycle;

public static class LifecycleLog
{
    public static string ArenaLoggerName(string arenaId)
    {
        var segment = (arenaId ?? string.Empty).Trim().Replace('.', '_');
        if (segment.Length == 0)
            return ArenaLogRouting.RootLoggerName;
        return $"{ArenaLogRouting.RootLoggerName}.{segment}";
    }

    internal static void LogTransitionDocument(string arenaId, ArenaLogRouting routing, string document)
    {
        ArenaState state;
        try
        {
            state = ArenaState.Parse(document);
        }
        catch (ArgumentException)
        {
            Emit(routing, ArenaLogRouting.RootLoggerName, LogLevel.Warning,
                $"unparseable arena state transition: {document}");
            return;
        }
        if (!string.Equals(state.Id, arenaId, StringComparison.Ordinal))
            return;
        LogTransition(state, routing);
    }

    internal static void LogTransition(ArenaState state, ArenaLogRouting routing)
    {
        var line = state.Faults.Count > 0 ? $"{state.State} | faults={state.Faults.Count}" : state.State;
        Emit(routing, ArenaLoggerName(state.Id),
            state.IsFaulted() ? LogLevel.Error : LogLevel.Information, line);
    }

    internal static void LogClosingSummaryDocument(string document, ArenaLogRouting routing)
    {
        ArenaState state;
        try
        {
            state = ArenaState.Parse(document);
        }
        catch (ArgumentException)
        {
            Emit(routing, ArenaLogRouting.RootLoggerName, LogLevel.Warning,
                $"unparseable arena closing state: {document}");
            return;
        }
        LogClosingSummary(state, routing);
    }

    internal static void LogClosingSummary(ArenaState state, ArenaLogRouting routing)
    {
        Emit(routing, ArenaLoggerName(state.Id), LogLevel.Information,
            $"closing summary | state={state.State} | faults={state.Faults.Count}");
    }

    private static void Emit(ArenaLogRouting routing, string loggerName, LogLevel level, string line)
    {
        var logger = routing.LoggerFor(loggerName);
        var message = routing.MessageFor(loggerName, line);
        logger.Log(level, 0, message, null, static (s, _) => s);
    }
}
